src/Bundle/Doctrine/ORM/ExpressionBuilder.php

```diff
@@ -195,70 +184,93 @@ private function adjustField(string $field): string
 
     private function resolveFieldByAddingJoins(string $field): string
     {
-        if (0 === substr_count($field, '.')) {
-            return $this->getFieldName($field);
-        }
+        $field = $this->getAbsolutePath($field);
+        $metadata = $this->getClassMetadataForRootField(explode('.', $field, 2)[0]);
 
-        $key = 0;
+        while (count($explodedField = explode('.', $field, 3)) === 3) {
+            [$rootField, $associationField, $remainder] = $explodedField;
 
-        $newAlias = $this->getAlias($key);
-        $fields = explode('.', $field);
-        foreach ($fields as $field) {
-            if ($key === count($fields) - 1) {
+            if (isset($metadata->embeddedClasses[$associationField])) {
                 break;
             }
 
-            if ($this->hasAlias($field)) {
-                $newAlias = $field;
-                ++$key;
+            $metadata = $this->queryBuilder->getEntityManager()->getClassMetadata(
+                $metadata->getAssociationMapping($associationField)['targetEntity']
+            );
+            $rootAndAssociationField = sprintf('%s.%s', $rootField, $associationField);
 
-                continue;
-            }
+            /** @var Join[] $joins */
+            $joins = array_merge([], ...array_values($this->queryBuilder->getDQLPart('join')));
```

Comment:
`array_merge([], ...array_values($this->queryBuilder->getDQLPart('join')));` maybe it could be some (even private static) helper function at the end of file with meaningful name like flattenArray?